feat: Upgrade to Intervention Image 3.x

eliot488995568 opened the pull request:
- Replace `Image` with `ImageInterface` for compatibility.
- Update processors to use new methods (e.g., `scaleDown`, `coverDown`).
- Remove deprecated methods and adapt to the 3.x API.
- Add `BackgroundColorProcessor` and remove `LimitColorsProcessor`.
- Support for watermark placement via configuration.
- Adapt README.md
- Add `ProgressiveSubscriber` for event handling.
- Extend `ImageEncoder` and `NoCacheImageRequestSubscriber` to support progressive encoding.
- Implement `ImageEncoderInterface`
- Remove `ProgressiveProcessor` in favor of event-based handling.
- Update `RequestEvent` and `ImageSavedEvent` for progressive flag support.
- Document changes in `UPGRADE.md`.

eliot488995568 committed:
feat!: Remove deprecated constant from ImageSavedEvent

eliot488995568 committed:
feat: Upgrade to Intervention Image 3.x

- Replace `Image` with `ImageInterface` for compatibility.
- Update processors to use new methods (e.g., `scaleDown`, `coverDown`).
- Remove deprecated methods and adapt to the 3.x API.
- Add `BackgroundColorProcessor` and remove `LimitColorsProcessor`.
- Support for watermark placement via configuration.

eliot488995568 committed:
feat: Update README for Intervention Image 3.x compatibility

- Update links and descriptions to align with Intervention Image 3.x API.

eliot488995568 commented:
@ambroisemaupate I have a slight doubt about the src/Processor/ProgressiveProcessor.php

ambroisemaupate commented:
Could you prepare a `UPGRADE.md` file to tell what will change for end-users?

- If I use this library in my PHP project
- If I use standalone Docker image

ambroisemaupate commented on src/Processor/ProgressiveProcessor.php:
```diff
@@ -19,12 +19,7 @@ public function process(Image $image, Request $request): void
-             * @see https://image.intervention.io/v3/basics/image-output
-             */
-            $image->interlace((bool) $process);
+            $image->encodeByMediaType(progressive: (bool) $process);
```
I think this logic must be moved from Processors to an Event Subscriber when Image is rendered or saved.

ambroisemaupate commented on src/Encoder/ImageEncoder.php:
```diff
@@ -16,14 +17,14 @@ class ImageEncoder
+    public function encode(ImageInterface $image, string $path, int $quality): EncodedImageInterface
     {
-        return $image->encode($this->getImageAllowedExtension($path), $quality);
+        return $image->encodeByExtension($this->getImageAllowedExtension($path), quality: $quality);
```
This is where you should add `progressive` param when requested.

ambroisemaupate commented on src/Listener/NoCacheImageRequestSubscriber.php:
```diff
@@ -37,7 +37,7 @@ public function onRequest(RequestEvent $requestEvent): void
             if ($nativeImage->isNextGen()) {
                 $response = new Response(
-                    (string) $image->encode($nativeImage->getNextGenExtension(), $requestEvent->getQuality()),
+                    (string) $image->encodeByExtension($nativeImage->getNextGenExtension(), $requestEvent->getQuality()),
```
We use `$image->encodeByExtension` and `$image->encode` at several places.

We should factorize this to use our own `ImageEncoder` which will support `quality` and `progressive` parameters

ambroisemaupate commented on src/Listener/NoCacheImageRequestSubscriber.php:
```diff
@@ -48,10 +48,10 @@ public function onRequest(RequestEvent $requestEvent): void
             } else {
                 $response = new Response(
-                    (string) $image->encode(null, $requestEvent->getQuality()),
+                    $image->encode()->toString(),
```
Same here

ambroisemaupate commented on src/Listener/NoCacheImageRequestSubscriber.php:
```diff
@@ -48,10 +48,10 @@ public function onRequest(RequestEvent $requestEvent): void
                     Response::HTTP_OK,
                     [
-                        'Content-Type' => $image->mime(),
+                        'Content-Type' => $image->encode()->mimetype(),
```
You should **not** reencode image just to get its mimetype!

Extract a locale variable here.

eliot488995568 committed:
test: Add ImageProcessingTest for fit image generation

eliot488995568 committed:
feat(tests): Add comprehensive unit tests for image processing

- Add tests for various image transformations
- Validate output using MD5 hash comparison.
- Document testing strategy and commands in README.
- Update GitHub Actions workflow to run unit tests.
- Add PHPUnit as a dependency in composer.json.
- Update RotateProcessor to allow custom background color for rotation.

eliot488995568 committed:
feat(tests): Add Unit Test and Functional Test

eliot488995568 committed:
feat: Introduce progressive image support

- Add `ProgressiveSubscriber` for event handling.
- Extend `ImageEncoder` and `NoCacheImageRequestSubscriber` to support progressive encoding.
- Remove `ProgressiveProcessor` in favor of event-based handling.
- Update `RequestEvent` and `ImageSavedEvent` for progressive flag support.
- Document changes in `UPGRADE.md`.

eliot488995568 committed:
Merge branch 'feat/upgrade_intervention_image' of github.com:eliot488995568/intervention-request into feat/add_unit_test

ambroisemaupate commented on src/Listener/ProgressiveSubscriber.php:
```diff
@@ -0,0 +1,75 @@
+
+final class ProgressiveSubscriber implements EventSubscriberInterface
+{
+    private bool $progressive;
+
+    public function __construct(bool $progressive)
+    {
+        $this->setProgressive($progressive);
+    }
```
```suggestion
    public function __construct(private bool $progressive = false)
    {
    }
```

ambroisemaupate commented on src/Listener/ProgressiveSubscriber.php:
```diff
@@ -0,0 +1,75 @@
+                $this->progressive
+            );
+            if (\is_numeric($progressive)) {
+                if (1 == $progressive) {
```
```suggestion
                if (1 === (int) $progressive) {
```

ambroisemaupate commented on src/Listener/ProgressiveSubscriber.php:
```diff
@@ -0,0 +1,75 @@
+            if (\is_numeric($progressive)) {
+                if (1 == $progressive) {
+                    $progressive = true;
+                } elseif (0 == $progressive) {
```
```suggestion
                } elseif (0 === (int) $progressive) {
```

ambroisemaupate commented on src/Listener/ProgressiveSubscriber.php:
```diff
@@ -0,0 +1,75 @@
+        ];
+    }
+
+    public function setProgressive(bool $progressive): ProgressiveSubscriber
+    {
+        $this->progressive = $progressive;
+
+        return $this;
+    }
```
```suggestion
```

No need to expose `setter`

ambroisemaupate commented on src/Listener/ProgressiveSubscriber.php:
```diff
@@ -0,0 +1,75 @@
+    {
+        if ($requestEvent->getRequest()->query->has('no_process')) {
+            // Do not alter progressive at image file save. but allow to post-process optimizer to use progressive info.
+            $this->setProgressive(false);
```
```suggestion
            $this->progressive = false;
```

ambroisemaupate commented on src/Listener/ProgressiveSubscriber.php:
```diff
@@ -0,0 +1,75 @@
+                }
+            }
+            if (\is_bool($progressive)) {
+                $this->setProgressive($progressive);
```
```suggestion
                $this->progressive = $progressive;
```

ambroisemaupate commented on src/InterventionRequest.php:
```diff
@@ -91,7 +100,8 @@ public function __construct(
             $chainProcessor,
-            $this->fileResolver
+            $this->fileResolver,
+            new ImageEncoder()
```
Share one instance of `ImageEncoder` with `FileCache`.

ambroisemaupate commented on src/Encoder/ImageEncoder.php:
```diff
@@ -4,8 +4,9 @@
+use Intervention\Image\Interfaces\EncodedImageInterface;
+use Intervention\Image\Interfaces\ImageInterface;
 
 class ImageEncoder
```
```suggestion
final class ImageEncoder implements ImageEncoderInterface
```

Create an interface with this class. And use the interface as dependency in your services constructors

eliot488995568 committed:
feat: Introduce `ImageEncoderInterface`

- Add `ImageEncoderInterface` and update `ImageEncoder` to implement it.
- Replace `ImageEncoder` usage with `ImageEncoderInterface` in constructors and dependencies.
- Refactor `ProgressiveSubscriber` to simplify progressive handling.
- Update `UPGRADE.md` with breaking changes and new interface requirements.

eliot488995568 committed:
Merge branch 'feat/upgrade_intervention_image' of github.com:eliot488995568/intervention-request into feat/add_unit_test

eliot488995568 committed:
refactor(tests): Simplify image processor tests with reusable methods and Updated README.

ambroisemaupate commented on src/Cache/FileCache.php:
You should pass the progressive parameter from RequestEvent to ImageSavedEvent.

In this file, the `onRequest` method will convert a `RequestEvent` to a `ImageSavedEvent`, both events carry `progressive` parameter:

- `$dispatcher->dispatch(new ImageSavedEvent($image, $cacheFile, $requestEvent->getQuality(), $requestEvent->isProgressive()));`

ambroisemaupate commented on UPGRADE.md:
```diff
@@ -0,0 +1,40 @@
+
+### Signature Changes
+
+* **The `Image` class now has its own dedicated interface `ImageInterface` and each function with the Image input or output signature now has its own interface.
```
Markdown and syntax issues:

```suggestion
* All classes and methods depending on `Image` class now depend on `ImageInterface`
```

ambroisemaupate commented on UPGRADE.md:
```diff
@@ -0,0 +1,40 @@
+* **The `Image` class now has its own dedicated interface `ImageInterface` and each function with the Image input or output signature now has its own interface.
+
+
+### Introduction of `ImageEncoderInterface`
+
+* **The `InterventionRequest` class now requires an implementation of `ImageEncoderInterface` to be passed as a constructor argument.**
+  Instead of passing `null`, ensure that you provide a valid `ImageEncoder` implementation when initializing the `InterventionRequest`.
```
This does not require a title, and you're missing some breaking changes:

```suggestion
* `InterventionRequest` class now requires an implementation of `ImageEncoderInterface` to be passed as a constructor argument.  Ensure that you provide a valid `ImageEncoderInterface` implementation such as our default `ImageEncoder`
* `Configuration::getDriver()` method returns `ImagickDriver|GdDriver` instead of a `string`
```

ambroisemaupate reviewed (changes requested):
Some changes, and I'm waiting your UnitTests and FunctionalTests with github action changes

eliot488995568 committed:
fix(dependencies): Downgrade PHPUnit to ^11.5 in composer.json

eliot488995568 committed:
fix(tests): Ensure temp directory is created for InterventionRequest test

eliot488995568 committed:
refactor(tests): Rename "testProcessedImageHash" to "processedImageHash"

eliot488995568 committed:
chore: Resolve threads

eliot488995568 committed:
Merge pull request #1 from eliot488995568/feat/add_unit_test



Unit Tests:
Each image processing strategy (Processor) is tested in isolation to ensure it behaves correctly under various input conditions.

Functional Tests:
Simulates full HTTP requests using the InterventionRequest handler to validate end-to-end image transformations based on URL parameters.

Validation:
All test responses are compared against expected results using MD5 checksums to ensure consistency and integrity of the generated images.

Documentation:
Test cases are clearly documented in the test classes and README for better maintainability and developer onboarding.

eliot488995568 committed:
chore: Fix threads
